### ISSUE TITLE
[Solved] I can't reproduce your results.

### ISSUE THREAD

**Issue by Hyeonjoong-Jang** [lmb-freiburg/autodispnet#2]
Hello, thank you for sharing a deployment code for your great work.

I finally managed to run your code, but I've got a question.

My environment is as below.
(Local)
Ubuntu 20.04 LTS 

(Nvidia Docker)
Ubuntu 16.04, GTX 1080 Ti, CUDA 8 
Python 3.5, Tensorflow-gpu 1.4.0, Cudnn 6, numpy 1.14, cmake 3.11.0

I successfully built lmbspecialops-eccv18 branch and used netdef_slim.
I could run the code with a pair of stereo images, but when I visualize the output .float3 file, it was too noisy.
![image](https://user-images.githubusercontent.com/55046566/86791162-35eb3d00-c0a4-11ea-9d9a-0ece2f752c68.png)

Did I miss something to run properly?
Could you please double check your deployment code? I'll wait for your response.

Thank you.

**Comment by tonmoy-saikia** [comment]
Did you download the weights?

**Comment by Hyeonjoong-Jang** [comment]
Yes, for the result above, I used 

CSS/training/00__FT3D__genotype/checkpoints/snapshot-150000

I also tried with CSS_kitti, but it showed a similar result.

Let me show my logs.

ssh://root@127.0.0.1:9984/root/anaconda3/envs/autodisp/bin/python -u /tmp/pycharm_project_161/nets/CSS/eval_CSS.py eval 1.png 2.png .
Using /root/lmbspecialops/build/lib/lmbspecialops.so
evolution <00__FT3D__genotype>: 00__FT3D__genotype        COMPLETE
Evo manager: complete
Evolution: /tmp/pycharm_project_161/nets/CSS/training/00__FT3D__genotype
// Eval scale is 1.0:
Graph is in train mode: False
===== Building net1 =====
Graph is in train mode: False
Tensor shapes are:
image_L : (1, 3, 320, 320)
image_L : (1, 3, 320, 320)
stem 0 : (1, 64, 160, 160)
stem 1 : (1, 128, 80, 80)
feature 0 : (1, 252, 40, 40)
encoder 0 : (1, 252, 40, 40)
encoder 1 : (1, 504, 20, 20)
encoder 2 : (1, 504, 20, 20)
encoder 3 : (1, 1008, 10, 10)
encoder 4 : (1, 1008, 10, 10)
encoder 5 : (1, 2016, 5, 5)
decoder 0 : (1, 1008, 10, 10)
decoder 1 : (1, 504, 20, 20)
decoder 2 : (1, 252, 40, 40)
decoder 3 : (1, 126, 80, 80)
===== Building net2 =====
Graph is in train mode: False
Tensor shapes are:
image_L : (1, 3, 320, 320)
image_L : (1, 3, 320, 320)
stem 0 : (1, 64, 160, 160)
stem 1 : (1, 128, 80, 80)
encoder 0 : (1, 252, 40, 40)
encoder 1 : (1, 252, 40, 40)
encoder 2 : (1, 504, 20, 20)
encoder 3 : (1, 504, 20, 20)
encoder 4 : (1, 1008, 10, 10)
encoder 5 : (1, 1008, 10, 10)
encoder 6 : (1, 2016, 5, 5)
decoder 0 : (1, 1008, 10, 10)
decoder 1 : (1, 504, 20, 20)
decoder 2 : (1, 252, 40, 40)
decoder 3 : (1, 126, 80, 80)
===== Building net3 =====
Graph is in train mode: False
Tensor shapes are:
image_L : (1, 3, 320, 320)
image_L : (1, 3, 320, 320)
stem 0 : (1, 64, 160, 160)
stem 1 : (1, 128, 80, 80)
encoder 0 : (1, 252, 40, 40)
encoder 1 : (1, 252, 40, 40)
encoder 2 : (1, 504, 20, 20)
encoder 3 : (1, 504, 20, 20)
encoder 4 : (1, 1008, 10, 10)
encoder 5 : (1, 1008, 10, 10)
encoder 6 : (1, 2016, 5, 5)
decoder 0 : (1, 1008, 10, 10)
decoder 1 : (1, 504, 20, 20)
decoder 2 : (1, 252, 40, 40)
decoder 3 : (1, 126, 80, 80)
2020-07-07 13:25:16.116437: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
2020-07-07 13:25:16.318484: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties: 
name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.683
pciBusID: 0000:17:00.0
totalMemory: 10.92GiB freeMemory: 10.77GiB
2020-07-07 13:25:16.318520: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:17:00.0, compute capability: 6.1)
loading /tmp/pycharm_project_161/nets/CSS/training/00__FT3D__genotype/checkpoints/snapshot-150000
restoring     global_step
INFO:tensorflow:Restoring parameters from /tmp/pycharm_project_161/nets/CSS/training/00__FT3D__genotype/checkpoints/snapshot-150000
Writing output to ./disp.float3

Process finished with exit code 0

**Comment by tonmoy-saikia** [comment]
Weird. 
For some reason, the weights are not restored. You should see it in the logs.
In your case, only the global step is restored. For me, it seems to be loading correctly.
```
...
restoring     tower_0/refine1/depth_wise_separable_conv_105/sequential_259/separable_conv2d_244/pointwise_kernel
restoring     tower_0/refine1/depth_wise_separable_conv_105/sequential_259/separable_conv2d_245/bias
restoring     tower_0/refine1/depth_wise_separable_conv_105/sequential_259/separable_conv2d_245/depthwise_kernel
restoring     tower_0/refine1/depth_wise_separable_conv_105/sequential_259/separable_conv2d_245/pointwise_kernel
restoring     tower_0/refine1/depth_wise_separable_conv_106/sequential_266/separable_conv2d_246/bias
restoring     tower_0/refine1/depth_wise_separable_conv_106/sequential_266/separable_conv2d_246/depthwise_kernel
restoring     tower_0/refine1/depth_wise_separable_conv_106/sequential_266/separable_conv2d_246/pointwise_kernel
restoring     tower_0/refine1/depth_wise_separable_conv_106/sequential_266/separable_conv2d_247/bias
...

```
Could you please check if the `var_names` are correctly loaded here: https://github.com/lmb-freiburg/netdef_slim/blob/346368231bd54623fdaf2334755d50b09e694f7e/tensorflow/tools/trainer/helpers.py#L173

**Comment by Hyeonjoong-Jang** [comment]
OK.
I can see that I only have restored global_step.
![image](https://user-images.githubusercontent.com/55046566/87265060-a9ee7080-c4fc-11ea-94d6-748561605055.png)
By the way, the version I downloaded uses "optimistic_restore" function instead of "optimistic_restore_getvar".
https://github.com/lmb-freiburg/netdef_slim/blob/346368231bd54623fdaf2334755d50b09e694f7e/tensorflow/tools/trainer/helpers.py#L274
Is it ok? Should I change these functions in the simpletrainer.py file?
![image](https://user-images.githubusercontent.com/55046566/87265345-87a92280-c4fd-11ea-880d-86b3e843990f.png)

**Comment by tonmoy-saikia** [comment]
The functions should work as-is. 
Could you check if your model variables are correct and let me know?

**Comment by Hyeonjoong-Jang** [comment]
I found that
https://github.com/lmb-freiburg/netdef_slim/blob/346368231bd54623fdaf2334755d50b09e694f7e/tensorflow/tools/trainer/helpers.py#L274
this code checks 'var' in tf.global_variables() is also in saved_shapes, and the only variable that satisfies this is 'global_step'.

Elements in saved_shapes have names like tower_0/ ...
while elements in tf.global_variables() have names like conv_re_lu/ ... as below.
![image](https://user-images.githubusercontent.com/55046566/87511682-a8fe4000-c6b0-11ea-9863-ae9311fc1a6f.png)

It seems like one of these two must be changed so that it can be consistent with the other.
How can I change this?

**Comment by tonmoy-saikia** [comment]
Hi, thanks for looking into this. Now I see your problem. I checked my setup and it seems that `tf.global_variables()`  is returning the right variable names for me. 
I suspect it has something to do with the tf version you are using.
The `tower` prefixes are added here:
https://github.com/lmb-freiburg/autodispnet/blob/master/operations/common.py#L75
The `name_scope` is only applied if `tf.Variable` is used and not  `tf.get_variable`. It could be that `tf.keras` used `tf.Variable` before (at the time of development, this code used tf 1.12). 

You could try downgrading to tf 1.12 or check if it works by changing `tf.name_scope` in https://github.com/lmb-freiburg/autodispnet/blob/master/operations/common.py#L75 to `tf.variable_scope`

**Comment by Hyeonjoong-Jang** [comment]
Hi, thank you for answering my questions. It feels like it is heading to the end.

I tried changing tf.name_scope to tf.variable_scope, and now variable names are starting with tower_0.
But in your log, variables have names with "tower_0/refine1", but mines don't have "refine1" after tower.

```
tower_0/depth_wise_separable_conv_125/sequential/separable_conv2d_1/pointwise_kernel
tower_0/depth_wise_separable_conv_125/sequential/separable_conv2d_1/bias
tower_0/dilated_separable_conv_41/sequential/separable_conv2d/depthwise_kernel
tower_0/dilated_separable_conv_41/sequential/separable_conv2d/pointwise_kernel
tower_0/dilated_separable_conv_41/sequential/separable_conv2d/bias
tower_0/conv2d_14/kernel
tower_0/conv2d_14/bias
global_step
restoring     global_step
restoring     tower_0/conv_re_lu/sequential/conv2d/bias
restoring     tower_0/conv_re_lu/sequential/conv2d/kernel
INFO:tensorflow:Restoring parameters from /tmp/pycharm_project_161/nets/CSS/training/00__FT3D__genotype/checkpoints/snapshot-150000
Writing output to ./disp.float3

Process finished with exit code 0
```



I am now using tf 1.4.0 because it is one of the requirements of lmbspecialops (https://github.com/lmb-freiburg/lmbspecialops/tree/eccv18).
Do you mean I should upgrade my tf version to 1.12, while keep using tf 1.4.0 for building lmbspecialops?

**Comment by tonmoy-saikia** [comment]
Ah sorry, I thought you were using tf 1.14 
No, please use tf 1.12.
 lmbspecialops should work with that as well.

**Comment by tonmoy-saikia** [comment]
> lmbspecialops should work with that as well.

You will need to recompile lmbspecialops, unfortunately.

**Comment by Hyeonjoong-Jang** [comment]
Oh yeah, tf 1.12 solved this problem right away.
Thank you.
